SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp

```diff
@@ -1599,6 +1645,46 @@ bool ExtractComponentSetsFromSchemaJson(const FString& SchemaJsonPath, TMap<uint
 			SAFE_TRYGETFIELD((*CompObject), Number, "componentId", ComponentId);
 
 			ComponentMap.Add(ComponentName, ComponentId);
+
+			const TArray<TSharedPtr<FJsonValue>>* FieldArray;
+			SAFE_TRYGETFIELD((*CompObject), Array, "fields", FieldArray);
+
+			uint32 TypeId = OutTypeIdToFieldIdsMap.Num();
+			if (FieldArray->Num() > 0)
+			{
+				SCHEMA_GEN_ERROR(OutComponentIdToTypeIdsMap.Contains(ComponentId),
+								 TEXT("The schema bundle contains duplicate component IDs with component %s."), *ComponentName, TypeId);
+				OutComponentIdToTypeIdsMap.Add(ComponentId, TypeId);
+
+				SCHEMA_GEN_ERROR(OutTypeIdToFieldIdsMap.Contains(TypeId),
+								 TEXT("Ran into a typeId collision with component %s, typeId: %d."), *ComponentName, TypeId);
+				OutTypeIdToFieldIdsMap.Add(TypeId);
+			}
+
+			for (const auto& ArrayValue : *FieldArray)
+			{
+				const TSharedPtr<FJsonObject>* ArrayObject;
+				SAFE_TRYGET(ArrayValue, Object, ArrayObject);
+
+				int32 FieldId;
+				SAFE_TRYGETFIELD((*ArrayObject), Number, "fieldId", FieldId);
+
+				SCHEMA_GEN_ERROR(OutTypeIdToFieldIdsMap[TypeId].FieldIds.Contains(FieldId),
+								 TEXT("The schema bundle contains duplicate fieldId: %d, component name: %s."), FieldId, *ComponentName);
+				OutTypeIdToFieldIdsMap[TypeId].FieldIds.Add(FieldId);
+			}
+
+			FString DataDefinition;
+			SAFE_TRYGETFIELD((*CompObject), String, "dataDefinition", DataDefinition);
+
+			if (!DataDefinition.IsEmpty())
```

Comment:
Switch this check to an XOR to catch when they are both empty?

Reply:
sorry, don't follow.
If both the dataDefinition and the fields are empty?
what should be xored?

Reply:
Sorry, confusing comment.
Check that either the data definition is set, or the field ids are set, but not neither nor both.

Reply:
Yeah, thought that may be it, however it's valid to have both empty if the component has no fields, like the tag components.

Reply:
Gotcha